packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart

```diff
@@ -52,8 +46,30 @@ Future<void> generateLocalizationsSyntheticPackage({
     if (isSyntheticL10nPackage == false) {
       return;
     }
+  } else if (!environment.useImplicitPubspecResolution) {
+    // --no-implicit-pubspec-resolution was passed, and synthetic-packages: true was not.
+    return;
+  }
+
+  if (!environment.useImplicitPubspecResolution) {
+    throwToolExit(
+      'Cannot generate a synthetic package when --no-implicit-pubspec-resolution is passed.\n'
+      '\n'
+      'Synthetic package output (package:flutter_gen) is deprecated: '
+      'https://flutter.dev/to/flutter-gen-deprecation. If you are seeing this '
+      'message either you have provided --no-implicit-pubspec-resolution, or '
+      'it is the default value (see flutter --verbose --help).',
+    );
   }
 
+  // Log a warning: synthetic-package: true (or implicit true) is deprecated.
+  environment.logger.printWarning(
+    'Synthetic package output (package:flutter_gen) is deprecated: '
+    'https://flutter.dev/to/flutter-gen-deprecation. In a future release '
+    'this value will default to `false`, and later support removed '
+    'entirely',
+  );
```

Comment:
Rhetorical question: what is "this value"?

Reply:
Clarified.